wire-ios-data-model/Source/Core Crypto/CoreCryptoKeyProvider.swift

```
@@ -23,90 +23,139 @@ import WireSystem
 
         do {
-            return try fetchCoreCryptoKey()
+            return try fetchScopedCoreCryptoKey()
         } catch {
             if createIfNeeded {
-                return try createCoreCryptoKey()
+                return try createScopedCoreCryptoKey()
             } else {
                 throw error
             }
         }
     }
 
     public func updateDatabaseKey(path: String) async throws {
-        if let oldKey = try? fetchCoreCryptoKey() {
-            let item = CoreCryptoKeychainItem()
-            let newKey = try KeychainManager.generateKey(numberOfBytes: 32)
-            try await coreCryptoKeyMigrationManager?.updateKey(path: path, oldKey: oldKey, newKey: newKey)
+        // We need to get the unscoped key because this is part of the migration to 4.3.0
+        // and scoped keys haven't been introduced before
+        if let oldKey = try? fetchUnscopedCoreCryptoKey() {
+            do {
+                WireLogger.coreCrypto.info("Updating core crypto key...", attributes: .safePublic)
+
+                let item = ScopedCoreCryptoKeychainItem(userID: userID)
+                let newKey = try KeychainManager.generateKey(numberOfBytes: 32)
+                try await coreCryptoKeyMigrationManager?.updateKey(path: path, oldKey: oldKey, newKey: newKey)
+
+                // In case another account needs to update it, we don't delete the old key because it's not scoped by
+                // account.
+                try KeychainManager.storeItem(item, value: newKey)
+            } catch {
+                WireLogger.coreCrypto.warn(
+                    "Failed to update core crypto key: \(String(describing: error))",
+                    attributes: .safePublic
+                )
+                throw error
+            }
+        }
+    }
+
+    public func migrateToScopedDatabaseKey(inactiveAccounts: Set<Account>) throws {
+        let scopedKeyExists = (try? fetchScopedCoreCryptoKey()) != nil
```

question: sanity check. Is it safe to treat an error as the key not existing?

---

not sure I understand your question, are you asking if it's safe to treat the key not existing as a condition to perform the migration ?

---

I mean that if an error is thrown it doesn't necessarily mean the key doesn't exist in the database. There could be some other error. I think the API of our KeychainManager isn't good here because it throws instead of returning `nil` when the data doesn't exist. So I think a better version of my question is?

If the error was thrown for another reason (not because the key doesn't exist), does that cause any logic issues?

---

I make a suggestion in my other comment related to this